exchanges/okcoin/okcoin_websocket.go

```diff
@@ -147,150 +238,236 @@ func (o *OKCoin) WsHandleData(respRaw []byte) error {
+			}
+			waitingSignatureLock.Unlock()
+			if o.Verbose {
+				log.Debugf(log.ExchangeSys,
+					o.Name+" - "+eventResponse.Event+" on channel: "+eventResponse.Channel)
 			}
-		}
-		if o.Verbose {
-			log.Debugln(log.ExchangeSys,
-				o.Name+" - "+eventResponse.Event+" on channel: "+eventResponse.Channel)
 		}
 	}
 	return nil
 }
 
-// StringToOrderStatus converts order status IDs to internal types
-func StringToOrderStatus(num int64) (order.Status, error) {
-	switch num {
-	case -2:
-		return order.Rejected, nil
-	case -1:
-		return order.Cancelled, nil
-	case 0:
-		return order.Active, nil
-	case 1:
-		return order.PartiallyFilled, nil
-	case 2:
-		return order.Filled, nil
-	case 3:
-		return order.New, nil
-	case 4:
-		return order.PendingCancel, nil
-	default:
-		return order.UnknownStatus, fmt.Errorf("%v not recognised as order status", num)
+func (o *Okcoin) wsProcessAdvancedAlgoOrder(respRaw []byte) error {
+	var resp WebsocketAdvancedAlgoOrder
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
+	}
+	o.Websocket.DataHandler <- resp
+	return nil
+}
+
+func (o *Okcoin) wsProcessAlgoOrder(respRaw []byte) error {
+	var resp WebsocketAlgoOrder
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
 	}
+	o.Websocket.DataHandler <- resp
+	return nil
 }
 
-func (o *OKCoin) wsProcessOrder(respRaw []byte) error {
-	var resp WebsocketSpotOrderResponse
+func (o *Okcoin) wsProcessOrders(respRaw []byte) error {
+	var resp WebsocketOrder
 	err := json.Unmarshal(respRaw, &resp)
 	if err != nil {
 		return err
 	}
-	for i := range resp.Data {
-		var oType order.Type
-		var oSide order.Side
-		var oStatus order.Status
-		oType, err = order.StringToOrderType(resp.Data[i].Type)
+	cp, err := currency.NewPairFromString(resp.Arg.InstrumentID)
+	if err != nil {
+		return err
+	}
+
+	algoOrder := make([]fill.Data, len(resp.Data))
+	for x := range resp.Data {
+		side, err := order.StringToOrderSide(resp.Data[x].PositionSide)
 		if err != nil {
-			o.Websocket.DataHandler <- order.ClassificationError{
-				Exchange: o.Name,
-				OrderID:  resp.Data[i].OrderID,
-				Err:      err,
-			}
+			return err
 		}
-		oSide, err = order.StringToOrderSide(resp.Data[i].Side)
+		algoOrder[x] = fill.Data{
+			ID:            resp.Data[x].OrderID,
+			Timestamp:     resp.Data[x].CreateTime.Time(),
+			Exchange:      o.Name,
+			AssetType:     asset.Spot,
+			CurrencyPair:  cp,
+			Side:          side,
+			OrderID:       resp.Data[x].OrderID,
+			ClientOrderID: resp.Data[x].ClientOrdID,
+			TradeID:       resp.Data[x].TradeID,
+			Price:         resp.Data[x].FillPrice,
+			Amount:        resp.Data[x].Size,
+		}
+	}
+	o.Websocket.DataHandler <- algoOrder
+	return nil
+}
+
+func (o *Okcoin) wsProcessAccount(respRaw []byte) error {
+	var resp WebsocketAccount
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
+	}
+	o.Websocket.DataHandler <- &resp
+	return nil
+}
+
+func (o *Okcoin) wsProcessOrderbook(respRaw []byte) error {
```

Comment:
I have recieved the following:
` exchange okcoin websocket error - websocket orderbook buffer error: update bid/ask targets cannot be nil`

The error could be more specific, which pair? which asset? Is the error a result of our processing? Or is it that okcoin is sending bad data?

Reply:
They have such push data below.
`{"arg":{"channel":"books","instId":"BTC-USD"},"action":"update","data":[{"asks":[],"bids":[],"ts":"1689510449762","checksum":1085589561}]}`

I added a check to skip such data. Thanks for pointing out.